platform/server/http/Router/index.ts

```diff
@@ -104,73 +125,103 @@ export class Router<V> {
 
-  get<P extends ObjectSetting<any>, Q extends ObjectSetting<any>>(
-    route: Route<P, Q>,
-    handler: RequestHandler<V, P, Q>
-  ) {
-    this.router.get(route.path, async (req, res) => {
-      let valid: { params: TypeOf<P>; query: TypeOf<P> };
-
-      // TODO Change this so we can get failures per type
-      try {
-        valid = KibanaRequest.validate(route, req);
-      } catch (e) {
-        res.status(400);
-        res.json({ error: e.message });
-        return;
-      }
-
-      const kibanaRequest = new KibanaRequest(req, valid.params, valid.query);
-
-      const value =
-        this.options.onRequest !== undefined
-          ? this.options.onRequest(kibanaRequest)
-          : {} as V;
-
-      try {
-        const response = await handler(value, kibanaRequest, responseFactory);
-
-        if (response instanceof KibanaResponse) {
-          res.status(response.status);
-
-          if (response.payload === undefined) {
-            res.send();
-          } else if (response.payload instanceof Error) {
-            // TODO Design an error format
-            res.json({ error: response.payload.message });
-          } else {
-            res.json(response.payload);
-          }
-        } else {
-          res.json(response);
-        }
-      } catch (e) {
-        // TODO Specifically handle `KibanaResponseError` and validation errors.
+  get<
+    P extends ObjectSetting<any>,
+    Q extends ObjectSetting<any>,
+    B extends ObjectSetting<any>
+  >(route: Route<P, Q, B>, handler: RequestHandler<V, P, Q, B>) {
+    this.router.get(
+      route.path,
+      async (req, res) => await this.handle(route, req, res, handler)
+    );
+  }
 
-        // Otherwise we default to something along the lines of
-        res.status(500).json({ error: e.message });
-      }
-    });
+  post<
+    P extends ObjectSetting<any>,
+    Q extends ObjectSetting<any>,
+    B extends ObjectSetting<any>
+  >(route: Route<P, Q, B>, handler: RequestHandler<V, P, Q, B>) {
+    this.router.post(
+      route.path,
+      async (req, res) => await this.handle(route, req, res, handler)
+    );
   }
 
-  post<P extends ObjectSetting<any>, Q extends ObjectSetting<any>>(
-    route: Route<P, Q>,
-    handler: RequestHandler<V, P, Q>
-  ) {
+  put<
+    P extends ObjectSetting<any>,
+    Q extends ObjectSetting<any>,
+    B extends ObjectSetting<any>
+  >(route: Route<P, Q, B>, handler: RequestHandler<V, P, Q, B>) {
     // TODO
   }
 
-  put<P extends ObjectSetting<any>, Q extends ObjectSetting<any>>(
-    route: Route<P, Q>,
-    handler: RequestHandler<V, P, Q>
-  ) {
+  delete<
+    P extends ObjectSetting<any>,
+    Q extends ObjectSetting<any>,
+    B extends ObjectSetting<any>
+  >(route: Route<P, Q, B>, handler: RequestHandler<V, P, Q, B>) {
     // TODO
   }
 
-  delete<P extends ObjectSetting<any>, Q extends ObjectSetting<any>>(
-    route: Route<P, Q>,
-    handler: RequestHandler<V, P, Q>
+  private async handle<
```

Comment:
Didn't try to generalize here, just moved duplicate code into single place and left generalization for later :)